### PR TITLE
Extend downward node labels support to more components

### LINKED ISSUE #9366: Feature: add downward-node-labels to Logstash / Kibana / Agent
**Is your feature request related to a problem? Please describe.**
I want to use node labels to set https://www.elastic.co/docs/reference/logstash/plugins/plugins-inputs-kafka#plugins-inputs-kafka-client_rack but downard-node-labels are only propogated for elasticsearch resources. 

**Describe the resource you would like to have implemented.**

Mirrors the existing Elasticsearch behavior: users can set the
eck.k8s.elastic.co/downward-node-labels annotation on the CR with a
comma-separated list of node labels, and the operator copies those labels
as annotations on the managed Pods. Pods can then read them via the
Downward API.

I primarily need this for Logstash but we might as well implement it for kibana / agent

### PR DESCRIPTION
**Pull request by pkoutsovasilis**
## What this PR does

For Elasticsearch, users can annotate a CR with `eck.k8s.elastic.co/downward-node-labels: "<label>"` and ECK copies the matching node label onto each pod as an annotation, making it consumable via the Kubernetes downward API (e.g. `fieldRef: metadata.annotations['topology.kubernetes.io/zone']`).

This PR extends the same feature to the remaining stack components: **Kibana, Logstash, APM Server, Beats, Agent, Elastic Maps Server, and Package Registry**.

Closes https://github.com/elastic/cloud-on-k8s/issues/9366

### Generic shared package (`pkg/controller/common/nodelabels`)

The Elasticsearch-specific implementation (`driver/shared/node_labels.go`, `validation/node_labels.go`) is extracted into a common package reusable by all controllers. Each component's controller calls `AnnotatePods` during its normal reconcile loop. Webhook validation of the annotation against the operator's `--exposed-node-labels` allow-list is wired consistently across all 8 component types.

### Shell-free init container (`cmd/waitforannotations`)

The previous bash/grep init container script broke on images that don't ship a shell or grep binary (e.g. wolfi-based APM Server). For the new components this is replaced by a new `wait-for-annotations` subcommand on the **operator binary itself**. The init container runs using the operator's own image (read from the `OPERATOR_IMAGE` env var injected by the Helm chart / OLM CSV, overridable via `--operator-image`), polls the downward-API-projected annotations file every 2 seconds until all expected annotation keys are present, then exits 0 - unblocking the main container.

This is a hard gate: the main container cannot start until the node label has been propagated to the pod annotation.

### Operator image injection

A new `OPERATOR_IMAGE` environment variable is injected into the operator pod (`statefulset.yaml`, `values.yaml`) so the init container builder can reference the correct image without pod introspection.

## Behaviour notes

- Errors from `AnnotatePods` propagate through the standard `reconciler.Results` path and surface in the CR's status.
- If `OperatorImage` cannot be determined, the operator logs a warning and returns a hard reconcile error only for CRs that use the annotation - CRs without it are unaffected.

## E2E test

A new e2e test (`test/e2e/apm/downward_node_labels_test.go`) covers the full happy path against APM Server backed by Elasticsearch:

1. The `wait-for-annotations` init container is injected and uses the operator image.
2. The init container terminates with exit code 0.
3. `topology.kubernetes.io/zone` is copied from the node as a pod annotation.
4. A `NODE_ZONE` env var declared via `fieldRef: metadata.annotations['topology.kubernetes.io/zone']` resolves to the correct zone value inside the running main container (verified with `printenv NODE_ZONE`).

The test targets APM Server only. The annotation-patching and init-container-building paths are shared across all components via `pkg/controller/common/nodelabels`, and each component already has dedicated pod-builder unit tests. In the spirit of not multiplying e2e suite time for coverage of logic that is already exercised by shared unit tests, a single representative component is sufficient here.

### CODE REVIEW

**Comment by pkoutsovasilis** [comment]
buildkite test this -f p=gke,t=TestDownwardNodeLabels,E2E_TAGS=apm

**Comment by pkoutsovasilis** [comment]
buildkite test this -f p=gke,t=TestDownwardNodeLabels,E2E_TAGS=apm

File: cmd/waitforannotations/main.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+		return fmt.Errorf("at least one --%s is required", AnnotationFlag)
+	}
+	timeout := viper.GetDuration(TimeoutFlag)
+	pollInterval := viper.GetDuration(PollIntervalFlag)
+
+	fmt.Printf("Waiting for annotations %v in %s\n", annotations, file)
+
+	start := time.Now()
+	ticker := time.NewTicker(pollInterval)
+	defer ticker.Stop()
```
[the remark this answers is not in the document]
Comment (pkoutsovasilis) [reply cmd/waitforannotations/main.go:92]: fixed in 3fc7c876a6a76d54f601a9d06d8e66ca19234576

File: test/e2e/test/apmserver/builder.go [hunk 2]
```diff
@@ -161,6 +161,25 @@ func (b Builder) WithHTTPCfg(cfg commonv1.HTTPConfig) Builder {
...
+			return b
+		}
+	}
+	containers = append(containers, corev1.Container{
+		Name: containerName,
+		Env:  envVars,
+	})
+	b.ApmServer.Spec.PodTemplate.Spec.Containers = containers
+	return b
+}
```
[the remark this answers is not in the document]
Comment (pkoutsovasilis) [reply test/e2e/test/apmserver/builder.go]: fixed in af0dd9ee0ce13779963f1d44aba4c5d9538da46e

File: pkg/controller/common/nodelabels/initcontainer.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+// which removes any dependency on the stack/component image having a shell or grep.
+//
+// operatorImage must be non-empty; an error is returned otherwise so callers fail loudly
+// rather than silently falling back to the stack image via PodTemplateBuilder.WithInitContainerDefaults.
+// Callers must also add the volume returned by DownwardAPIVolume to the Pod.
+func WaitForAnnotationsInitContainer(operatorImage string, expectedAnnotations []string) (corev1.Container, error) {
+	if operatorImage == "" {
+		return corev1.Container{}, errors.New("operator image is required to build the wait-for-annotations init container; " +
+			"set --operator-image or ensure the operator can introspect its own Pod")
+	}
```
[the remark this answers is not in the document]
Comment (pkoutsovasilis) [reply pkg/controller/common/nodelabels/initcontainer.go]: fixed in c027c72ebcdaca80dabdb614c73c0e3ee50525e4

File: pkg/controller/common/nodelabels/exposed.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	"regexp"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	commonv1 "github.com/elastic/cloud-on-k8s/v3/pkg/apis/common/v1"
+)
+
+// NotAllowedNodesLabelMsg is returned when a node label requested via the downward-node-labels
+// annotation is not allowed by the operator's exposed-node-labels policy.
+const NotAllowedNodesLabelMsg = "Node label not in the exposed node labels list"
```
Comment (barkbay) [inline pkg/controller/common/nodelabels/exposed.go:18]: This is duplicated in https://github.com/elastic/cloud-on-k8s/blob/4ec6156fc9b2034f7b057b59aea344a961a0f09d/pkg/controller/elasticsearch/validation/validations.go#L47

File: pkg/telemetry/telemetry.go [hunk 5]
```diff
@@ -397,10 +432,27 @@ func beatStats(k8sClient k8s.Client, managedNamespaces []string) (string, any, e
...
-	return "beats", stats, nil
+	if resourcesWithDownwardLabels == 0 {
+		return "beats", stats, nil
+	}
+
+	ret := make(map[string]any, len(stats)+1)
+	for k, v := range stats {
+		ret[k] = v
+	}
+	ret["downward_node_labels"] = map[string]any{
```
Comment (barkbay) [inline pkg/telemetry/telemetry.go]: Can we reuse `downwardNodeLabelsStats` as for the other resources?

File: docs/reference/eck-configuration-flags.md [hunk 6]
```diff
@@ -29,7 +29,7 @@ The following table lists and describes all the available configuration flags fo
...
 | `enable-webhook` | `false` | Enables a validating webhook server in the operator process. |
 | `enforce-rbac-on-refs` | `false` | Enables restrictions on cross-namespace resource association through RBAC. |
-| `exposed-node-labels` | `""` | List of Kubernetes node labels which are allowed to be copied as annotations on the Elasticsearch Pods. Check [Topology spread constraints and availability zone awareness](docs-content://deploy-manage/deploy/cloud-on-k8s/advanced-elasticsearch-node-scheduling.md#k8s-availability-zone-awareness) for more details. |
+| `exposed-node-labels` | `""` | Comma-separated list of Kubernetes node labels which are allowed to be copied as annotations on Pods (supported for Elasticsearch, Kibana, APM Server, Logstash, Beats, Agent, Elastic Maps Server, and Package Registry). Check [Topology spread constraints and availability zone awareness](docs-content://deploy-manage/deploy/cloud-on-k8s/advanced-elasticsearch-node-scheduling.md#k8s-availability-zone-awareness) for more details. |
```
Comment (barkbay) [inline docs/reference/eck-configuration-flags.md:33]: This still links to the Elasticsearch-only availability-zone-awareness page while claiming the flag covers eight components. Should we create an issue to track the doc update?

File: pkg/controller/agent/controller.go [hunk 7]
```diff
@@ -231,9 +231,8 @@ func (r *ReconcileAgent) doReconcile(ctx context.Context, agent agentv1alpha1.Ag
 
 func (r *ReconcileAgent) validate(ctx context.Context, agent agentv1alpha1.Agent) error {
 	defer tracing.Span(&ctx)()
-
-	// Run create validations only as update validations require old object which we don't have here.
-	warnings, err := agentv1alpha1.Validate(&agent, nil)
+	// Run create validations only: update validations require the old object which the reconciler does not have.
+	warnings, err := validateAgent(ctx, &agent, nil, r.licenseChecker, r.ExposedNodeLabels)
```
Comment (barkbay) [inline pkg/controller/agent/controller.go:236]: This moves the reconcile path from `agentv1alpha1.Validate` to `validateAgent`, which pulls in the
licence-gated `validClientAuthentication` (webhook.go:63). That looks deliberate, and it does make
Agent consistent with Elasticsearch, whose reconciler has run licence-gated validations for a while
(`elasticsearch_controller.go:268` calls the same `ValidateElasticsearch` as the webhook, licence
checker included). I mostly want to confirm one consequence is intended before it lands.

There are three situations here and they do not all want the same answer:

| Situation | before | after |
|---|---|---|
| create or update, webhook on, Basic licence | rejected at admission | unchanged |
| create or update, webhook off, Basic licence | accepted, driver warns and degrades | reconcile fails |
| existing Agent, licence lapses, no write | reconciles, driver warns and degrades | reconcile fails |

Row 2 is a real inconsistency and this fixes it. The "fallback in case webhook is disabled" comment
right above the call argues for exactly that, so no concern from me.

Row 3 is the one worth a second look. That Agent was created while the licence was valid. After
this change a lapse stops its reconciliation entirely, so transport certificates stop rotating and
version or config changes are not applied until the licence is restored.

The concrete symptom is that it makes the degradation branch at `driver.go:279` unreachable:

- `internalReconcile` has exactly one caller (`controller.go:218`), which runs after `r.validate`
  (`controller.go:206`).
- The warning fires on `specEnabled && !enterpriseEnabled && !hasAnnotation`. The new check errors
  on `specEnabled && !enterpriseEnabled`, which is strictly weaker, so it always fires first.
- The only path left is a transient where the licence checker errors during validation, which fails
  open by design, and then succeeds inside the driver.

So we would be asserting two different things about the same state: validation says reject, the
driver says warn and carry on. `grep "ineffective because an Enterprise license"` returns only that
one line, so ES never had this tension.

<details>
<summary>Test that shows the delta, if useful</summary>

```go
func fleetServerWithClientAuth() *agentv1alpha1.Agent {
	return &agentv1alpha1.Agent{
		ObjectMeta: metav1.ObjectMeta{Name: "fleet-server", Namespace: "default"},
		Spec: agentv1alpha1.AgentSpec{
			Version:            "9.5.0",
			Mode:               agentv1alpha1.AgentFleetMode,
			FleetServerEnabled: true,
			PolicyID:           "eck-fleet-server",
			Deployment:         &agentv1alpha1.DeploymentSpec{},
			HTTP: commonv1.HTTPConfigWithClientOptions{
				TLS: commonv1.TLSWithClientOptions{
					Client: commonv1.ClientOptions{Authentication: true},
				},
			},
		},
	}
}

func TestAgentValidationDelta(t *testing.T) {
	basic := commonlicense.MockLicenseChecker{EnterpriseEnabled: false}

	// what the reconciler called before: passes on Basic, driver then warns and degrades
	_, err := agentv1alpha1.Validate(fleetServerWithClientAuth(), nil)
	require.NoError(t, err)

	// what the reconciler calls now: reconcile aborts before the driver runs
	_, err = validateAgent(context.Background(), fleetServerWithClientAuth(), nil, basic, nil)
	require.Error(t, err)
	assert.Contains(t, err.Error(), "client certificate authentication requires an enterprise license")
}
```

</details>

File: pkg/controller/packageregistry/pod.go [hunk 8]
```diff
@@ -128,7 +128,12 @@ func newPodSpec(epr eprv1alpha1.PackageRegistry, configHash string, meta metadat
 	// Add HTTP certificates volume if TLS is enabled
 	builder = withHTTPCertsVolume(builder, epr)
 
-	return builder.PodTemplate, nil
+	builder, err = commonnodelabels.MaybeAddWaitForAnnotationsInitContainer(builder, &epr, operatorImage)
```
Comment (barkbay) [inline pkg/controller/packageregistry/pod.go:123]: This lands after the `WithContainersSecurityContext` call on line 107, so I think the wait container ends up as the only container in the pod without a `securityContext`. That setter writes into the containers that exist when it runs, and its doc comment asks to be called "once all the Containers and InitContainers have been set". `WithInitContainerDefaults` propagates image, mounts, resources and env, but not security context, so nothing fills it in later.

It matters because EPR's main container is currently compliant with the PodSecurity `restricted` profile (`allowPrivilegeEscalation: false`, `capabilities.drop: [ALL]`, and `runAsNonRoot` on 9.3.0+). In a namespace with `pod-security.kubernetes.io/enforce=restricted`, any EPR pod carrying the annotation is then refused by admission:

```
Warning  FailedCreate  ... violates PodSecurity "restricted:latest":
  allowPrivilegeEscalation != false (container "elastic-internal-wait-for-node-labels"...),
  unrestricted capabilities (...), runAsNonRoot != true (...)
```

Note that the same pattern exists in `pkg/controller/kibana/pod.go`. APM, Maps, Logstash, Beat and Agent are unaffected, since they only use `WithPodSecurityContext`, which is pod-level and applies whenever a container is added.

Worth noting that `TestNewPodSpec_DownwardNodeLabels` passes `setDefaultSecurityContext=false`, so it does not exercise this path. A `true` case asserting the init container has a `SecurityContext` would lock it down.

File: cmd/waitforannotations/main.go [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+)
+
+const (
+	FileFlag         = "file"
+	AnnotationFlag   = "annotation"
+	TimeoutFlag      = "timeout"
```
Comment (barkbay) [inline cmd/waitforannotations/main.go:24]: I think `--timeout` and `--poll-interval` are never passed by the operator: `WaitForAnnotationsInitContainer` only sets `wait-for-annotations`, `--file=` and `--annotation=`. So in production the timeout is always 0, meaning wait forever, and this flag is unreachable through the operator.

Is it meant to be user-facing? Maybe it does work today via the env binding?

```bash
      $ TIMEOUT=2s ./elastic-operator wait-for-annotations --file=./annotations --annotation=topology.kubernetes.io/zone
      Waiting for annotations [topology.kubernetes.io/zone] in ./annotations
      Error: timed out after 2s waiting for annotations [...]
```

Which a user could reach with:

```yaml
  podTemplate:
    spec:
      initContainers:
      - name: elastic-internal-wait-for-node-labels
        env:
        - name: TIMEOUT
          value: 15m
```

Their env survives the merge, since `Defaulter.WithEnv` only adds names that are not already present and this container declares no env.

If that is the intent, it is worth documenting?

This also creates an asymmetry with Elasticsearch, and ES is the less configurable side. So IIUC documenting a timeout would indeed be true for all apps except ES?

File: pkg/controller/common/nodelabels/initcontainer.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+		"wait-for-annotations",
+		"--file=" + DownwardAPIVolume().AnnotationsFilePath(),
+	}
+	for _, a := range expectedAnnotations {
+		cmd = append(cmd, "--annotation="+a)
+	}
+
+	return corev1.Container{
+		Name:    WaitForAnnotationsContainerName,
+		Image:   operatorImage,
```
Comment (barkbay) [inline pkg/controller/common/nodelabels/initcontainer.go]: Something I did not realize immediately: putting the operator image in the pod template couples every managed pod that uses this feature to the operator's own version, so every ECK upgrade rolls all of them, including patch releases where nothing about the component changed.

Comment (pkoutsovasilis) [reply pkg/controller/common/nodelabels/initcontainer.go]: Ok amazing catch, I didn't connect the dots here... IMO this is a blocker for this PR, this isn't acceptable. I am investigating potential solutions brb

File: pkg/controller/apmserver/webhook.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package apmserver
```
Comment (barkbay) [inline pkg/controller/apmserver/webhook.go:5]: The PR adds five new files that looks structurally identical:

- `pkg/controller/apmserver/webhook.go`
- `pkg/controller/beat/webhook.go`
- `pkg/controller/kibana/webhook.go`
- `pkg/controller/maps/webhook.go`
- `pkg/controller/packageregistry/webhook.go`

I wonder if we can simplify things using a "decorator"?

```go
func WithAnnotationCheck[T metav1.Object](
	gk schema.GroupKind,
	exposedNodeLabels NodeLabels,
	validate func(obj T, old T) (admission.Warnings, error),
) func(obj T, old T) (admission.Warnings, error) {
	return func(obj T, old T) (admission.Warnings, error) {
		warnings, err := validate(obj, old)
		if err != nil {
			return warnings, err
		}
		if errs := ValidateAnnotation(obj.GetAnnotations(), exposedNodeLabels); len(errs) > 0 {
			return warnings, apierrors.NewInvalid(gk, obj.GetName(), errs)
		}
		return warnings, nil
	}
}
```

File: pkg/controller/common/nodelabels/initcontainer.go (hunk 10 above)
Comment (barkbay) [reply pkg/controller/common/nodelabels/initcontainer.go]: One alternative would be maybe to use `container.ImageRepository(container.ElasticsearchImage, ver)` (where ver is the component's version) as the init container image, and replace the operator binary call with a bash loop like the one in ES. The init container image would then track the component version instead of the operator version. While using the ES image in this context for other applications may sound awkward, we know it would work and the user can always override the image in the PodTemplate.

File: pkg/controller/common/nodelabels/exposed.go (hunk 4 above)
Comment (pkoutsovasilis) [reply pkg/controller/common/nodelabels/exposed.go:18]: deduped in edd66b9793984045165d0a36e9152d666071d6e0

File: pkg/telemetry/telemetry.go (hunk 5 above)
Comment (pkoutsovasilis) [reply pkg/telemetry/telemetry.go]: initially I tried to make this be totally based on a struct, however the potential unknown types of beats made my plans obsolete. But for this particular one you are right we can, changed in  f300637d9fa601413dd6cace79c5515cd0e4ba39

File: pkg/controller/agent/controller.go (hunk 7 above)
Comment (pkoutsovasilis) [reply pkg/controller/agent/controller.go:236]: good catch! I did align row 3 with the behaviour on main in cd040d4abd482991c2dff0f2a1e55efe0f01e48f and added some unit-test to guarantee that I won't forget about this the next time

File: pkg/controller/packageregistry/pod.go (hunk 8 above)
Comment (pkoutsovasilis) [reply pkg/controller/packageregistry/pod.go:123]: more than a valid point, fixed in a6d9aae374bca7c5bc413735584fe0432c09a528

File: cmd/waitforannotations/main.go (hunk 9 above)
Comment (pkoutsovasilis) [reply cmd/waitforannotations/main.go:24]: In my thinking, I introduced them as an escape hatch in case of user-reported issues (e.g. a stuck init container, or something else that may come up 🙂 ), but I don't believe they should be user-facing features. Would you like to remove the flags for now? we can always add them back with proper documentation if there's a concrete use case.

File: pkg/controller/common/nodelabels/initcontainer.go (hunk 10 above)
Comment (pkoutsovasilis) [reply pkg/controller/common/nodelabels/initcontainer.go]: In b9ebfa4 I took a different approach; before computing the template hash, the operator replaces the ECK-managed init container's image with a stable identity string (`managed:v1`). This means operator upgrades, which change the image tag, no longer affect the hash and don't roll pods. The identity includes a version token (`v1`) that can be bumped when the container's command semantics actually change and a pod roll is warranted. User-supplied images (via podTemplate) are left untouched and participate in the hash normally. Let me know what you think of that

File: pkg/controller/apmserver/webhook.go (hunk 11 above)
Comment (pkoutsovasilis) [reply pkg/controller/apmserver/webhook.go:5]: The decorator is the proper solution for the webhook path, but the reason I chose to go with new files instead is that the reconciler of each component also needs to have such a validation (e.g. to catch cases where the webhook is disabled). If we go with the decorator, we'd need a second call site inside the reconciler which means two places to check whenever the validation logic changes. Thus we will get two detached validation flows, where changes in one don't affect the other. Do you want me to go down the path of the decorator?

File: pkg/controller/common/nodelabels/initcontainer.go (hunk 10 above)
Comment (pkoutsovasilis) [reply pkg/controller/common/nodelabels/initcontainer.go]: Also, I did manually test the following scenarios:
- Real workload changes still cause an update and pick up the new operator image.
- User-overridden init container image remain part of the hash.
- Bumping HashVersion causes an intentional rollout.

File: docs/reference/eck-configuration-flags.md (hunk 6 above)
Comment (pkoutsovasilis) [reply docs/reference/eck-configuration-flags.md:33]: Yes I will create one, as soon as we resolve [this](https://github.com/elastic/cloud-on-k8s/pull/9583#discussion_r3689098517). Still this is a big no-go for me, if we aren't happy with the solution 🙂

**Comment by pkoutsovasilis** [comment]
buildkite test this -f p=gke -m E2E_TAGS=e2e

File: pkg/controller/kibana/pod.go [hunk 12]
```diff
@@ -213,7 +214,17 @@ func NewPodTemplateSpec(
...
-	return builder.WithInitContainerDefaults(additionalInitEnvVars...).PodTemplate, nil
+	builder, err = commonnodelabels.MaybeAddWaitForAnnotationsInitContainer(builder, &kb, operatorImage)
+	if err != nil {
+		return corev1.PodTemplateSpec{}, err
+	}
+
+	// WithContainersSecurityContext must be called after all containers and init containers have been added.
+	result := builder.WithInitContainerDefaults(additionalInitEnvVars...)
+	if canEnableSecurityContext {
+		result = result.WithContainersSecurityContext(defaultSecurityContext)
```
Comment (barkbay) [inline pkg/controller/kibana/pod.go:221]: Not checked on my side, but Claude tells me that moving this to the end also brings the stack monitoring sidecars into scope, since `stackmon.WithMonitoring` runs at line 213 and `pkg/controller/common/stackmon/sidecar.go` never sets a container SecurityContext. Rendering a Kibana with metrics and logs monitoring and no downward-node-labels annotation, metricbeat and filebeat go from securityContext=nil to the full defaultSecurityContext including `readOnlyRootFilesystem: true`, so those Kibana pods roll on upgrade to this release even when the feature is unused.

ES adds its own monitoring sidecars after WithContainersSecurityContext (nodespec/podspec.go:149 then :166), so its sidecars keep a nil one. The keystore init container is unaffected since it sets its own. EPR is fine, it has no sidecars.

Comment (pkoutsovasilis) [reply pkg/controller/kibana/pod.go:221]: Good catch, fixed in 927fcba73c55937cfa63f8c69a09e87b7765f20e by reordering the calls so WithContainersSecurityContext runs before stackmon.WithMonitoring and added a regression test in TestMonitoringSidecarsSecurityContext that fails if the monitoring sidecars have a security context set.

The above said, I am wondering should `canEnableSecurityContext` actually apply to the monitoring sidecar containers? If the intent is to harden all containers in the pod, the sidecars arguably ought to be included too. The counterargument is that metricbeat/filebeat are not tested under `readOnlyRootFilesystem: true` and silently applying it could break them. 🤔

File: pkg/controller/common/nodelabels/nodelabels.go [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+// initContainerHashIdentity returns the identity string to record in initcontainer.HashAnnotation.
+// If the pod template already contains a container named initcontainer.ContainerName with a non-empty
+// image, the user explicitly supplied it. Its image participates in the workload hash via the
+// pod spec directly (NormalizeTemplateForHash leaves it untouched), so the annotation only needs
+// to mark it as user-supplied ("user"). Otherwise ECK provides the image and operator
+// patch-upgrades must not roll pods ("managed:<version>").
+func initContainerHashIdentity(template corev1.PodTemplateSpec) string {
+	for _, c := range template.Spec.InitContainers {
+		if c.Name == initcontainer.ContainerName && c.Image != "" {
+			return "user"
```
Comment (barkbay) [inline pkg/controller/common/nodelabels/nodelabels.go]: `NormalizeTemplateForHash` only ever compares the identity against `"managed:"+HashVersion`, and it early-returns when the annotation is absent, so `user` and "no annotation" take the same path and the image is hashed as-is either way. Could we just not write the annotation when the user supplied the image? Also, on a bump the version ends up in the hashed template twice, once as the annotation value and once as the substituted image via `c.Image = identity`, so either alone would move the hash. Separately, the comment "the annotation itself records the image" looks stale, the value is the literal "user".

**Review by barkbay — COMMENTED** [review]
Admittedly a bit far-fetched, but I just realized another edge case with pulling the operator image: how do we know the `imagePullSecrets` that work for the application image are also valid for the operator one ECK never writes them into managed pods, so the pod relies on whatever the user set in `podTemplate.spec.imagePullSecrets`, which was chosen for the stack registry.

File: pkg/controller/common/nodelabels/nodelabels.go (hunk 13 above)
Comment (pkoutsovasilis) [reply pkg/controller/common/nodelabels/nodelabels.go]: Good point. In 40e42100593421d49cecf3f79d45ac49a1f5e7a9 I removed the "user" marker entirely. When the user supplies the init-container image, the managed annotation is now absent and the image participates in the Pod-spec hash normally. Any stale managed annotation is explicitly removed.

I also separated the versioned annotation value from the version-independent image placeholder, so HashVersion contributes to the hash only once. The stale comments and tests have been updated accordingly.

**Comment by pkoutsovasilis** [comment]
> Admittedly a bit far-fetched, but I just realized another edge case with pulling the operator image: how do we know the `imagePullSecrets` that work for the application image are also valid for the operator one ECK never writes them into managed pods, so the pod relies on whatever the user set in `podTemplate.spec.imagePullSecrets`, which was chosen for the stack registry.

Good point, I was aware of this during manual testing but I thought it was an acceptable trade-off that could be covered by the updated documentation for this feature. Users who need to pull from a private registry would need to either add the operator image's pull secret in `podTemplate.spec.imagePullSecrets` or configure it via the pod's service account. Is documenting that sufficient, or do you think we need to handle this in code?

File: pkg/controller/agent/controller.go [hunk 14]
```diff
@@ -231,13 +232,14 @@ func (r *ReconcileAgent) doReconcile(ctx context.Context, agent agentv1alpha1.Ag
...
 	if err != nil {
-		logconf.FromContext(ctx).Error(err, "Validation failed")
 		k8s.MaybeEmitErrorEvent(r.recorder, err, &agent, events.EventReasonValidation, events.EventActionValidation, err.Error())
-		return tracing.CaptureError(ctx, err)
+		if _, ok := errors.AsType[*licenseValidationError](err); !ok {
+			logconf.FromContext(ctx).Error(err, "Validation failed")
+			return tracing.CaptureError(ctx, err)
```
[the remark this answers is not in the document]
Comment (pkoutsovasilis) [reply pkg/controller/agent/controller.go:241]: this is inaccurate errors.AsType is part of the standard library and you can tell from the CI run this compiles just fine

File: pkg/controller/common/nodelabels/initcontainer.go (hunk 10 above)
Comment (barkbay) [reply pkg/controller/common/nodelabels/initcontainer.go]: If we fix a vuln in the binary or a dep (e.g. Go runtime), do we bump v1?

Comment (pkoutsovasilis) [reply pkg/controller/common/nodelabels/initcontainer.go]: Good question. Although the init container ships with the operator binary, I'm inclined to say no. HashVersion should only be bumped when a fix is specific to the wait-for-annotations subcommand itself (e.g. a bug in the polling logic or annotation parsing). A generic Go runtime or dependency vuln fix in the operator doesn't warrant rolling every managed pod. Also, whenever a pod is recreated for any other reason (config change, rolling restart, ~~node drain~~), it will pick up the operator image current at that time and the exposure is also narrow; the init container exits as soon as all annotations are present and doesn't stay running. Does that make sense?

Comment (barkbay) [reply pkg/controller/common/nodelabels/initcontainer.go]: > node drain

How would an eviction update the Pod Spec so the new Pod picks up the current operator image?

Comment (pkoutsovasilis) [reply pkg/controller/common/nodelabels/initcontainer.go]: you are right node drain won't do that, but everything else in there will

**Comment by pebrc** [comment]
Sorry for being late here. https://github.com/elastic/cloud-on-k8s/issues/9366#issuecomment-5240942298 points in  the right direction in my opinion. Let's see if KEP-4742 can cover most use cases. I have concerns with the proposed implementation. While I find the init container stable tagging clever it raises the question if the complexity is worth it if an upstream feature gets us very close to what we want to achieve here. I would at least pause until we get a response to @barkbay 's question from the user.

**Comment by pkoutsovasilis** [comment]
I am gonna close this one since [KEP-4742](https://github.com/kubernetes/enhancements/issues/4742) exists and will graduate to GA hopefully soon. I will open a follow-up documentation PR to capture that the ECK downward feature is only supported for elasticsearch and for other components users could try and utilise KEP-4742
